### ISSUE TITLE
Zero-arity function misbehaving

### ISSUE THREAD

**Issue by anderseknert** [open-policy-agent/opa#6314]
```rego
package p

foo() := "foo" {
    false
}

foo() := input.foo

bar := foo()
```

```
opa eval -f pretty -d p.rego 'data.p.foo'
undefined
```

Changing `foo() := input.foo` -> `foo() := 1`

```
opa eval -f pretty -d p.rego 'data.p.foo'
1
```

Changing `foo() := input.foo` -> `foo() := "bar"`

```
opa eval -f pretty -d p.rego 'data.p.foo'
1 error occurred: p.rego:9: rego_type_error: undefined function data.p.foo
```

It's Friday, so you'd think `foo` would be longing for a `bar`, but clearly the type checker wants its TPS report done by Saturday.

**Comment by ashutosh-narkar** [comment]
@anderseknert is this something we can address as part of https://github.com/open-policy-agent/opa/issues/6315? If yes, we can close this and track https://github.com/open-policy-agent/opa/issues/6315.

**Comment by anderseknert** [comment]
Well, if the outcome of that is that using zero-arity functions renders errors, I guess this one is _done_ already, and we'll just need to fix the cases that currently works 😄 Kidding aside though, I think it's good to track bugs separately from features or larger changes. If this bug is fixed once #6315 is deployed, that'd be a good time to close this one I think. But if you'd rater keep track of this in that ticket, feel free to do so.

**Comment by anderseknert** [comment]
Adding as a test case for #6315 as you suggested @ashutosh-narkar 👍
